File: leetcode/lc-us/py/2536.py
from typing import List


class Solution:
    def rangeAddQueries(self, n: int, queries: List[List[int]]) -> List[List[int]]:
        diff = [[0] * (n + 2) for _ in range(n + 2)]

        for q in queries:
            x1, y1, x2, y2 = q[0] + 1, q[1] + 1, q[2] + 1, q[3] + 1
            diff[x1][y1] += 1
            diff[x1][y2 + 1] -= 1
            diff[x2 + 1][y1] -= 1
            diff[x2 + 1][y2 + 1] += 1
        res = [[0] * n for _ in range(n)]
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                diff[i][j] += diff[i - 1][j] + diff[i][j - 1] - diff[i - 1][j - 1]
                res[i - 1][j - 1] = diff[i][j]
        return res

File: leetcode/lc-us/cpp/2536.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> rangeAddQueries(int n, vector<vector<int>> &queries) {
        vector<vector<int>> res(n, vector<int>(n));
        int diff[n + 2][n + 2];
        memset(diff, 0, sizeof diff);
        for (auto &q: queries) {
            int x1 = q[0] + 1, y1 = q[1] + 1, x2 = q[2] + 1, y2 = q[3] + 1;
            diff[x1][y1]++;
            diff[x2 + 1][y1]--;
            diff[x1][y2 + 1]--;
            diff[x2 + 1][y2 + 1]++;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                diff[i][j] += diff[i - 1][j] + diff[i][j - 1] - diff[i - 1][j - 1];
                res[i - 1][j - 1] = diff[i][j];
            }
        }
        return res;
    }
};